### PR TITLE
807 restructuring of applications folder and verification setups

### PR DESCRIPTION
**Pull request by jhabriel**
## Proposed changes

This PR introduces the following changes (see also #807):

1. The applications folder is now divided into two subdirectories: `building_blocks` and `complete_setups`.
2. The module `verification_utils.py` was moved to `building_blocks`. Data-saving-related functionality was moved to the core.
3. Mandel and Terzaghi setups were moved to `complete_setups`, whereas setups based on manufactured solutions were moved to `tests.functional.setups`. 
1. Mandel's and Terzaghi's setup now retrieves domain sizes from `domain.bounding_box`.
4. Terzaghi's setup has been modified to allow for scaled units via `pp.Units`. A new test has been written to check the incorporated functionality.
5. A model glossary functionality has been included. The glossary aims at facilitating easy access to (almost) all known methods/variables that are included in the mixin classes as part of the new model framework. 

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply_

- [ ] Minor change (e.g., dependency bumps, broken links, etc).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [x] Maintenance (e.g., improve logic and performance, remove obsolete code, etc).
- [ ] Other: 

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.	
- [x] Static typing is included in the update.
- [x] This PR does not duplicated existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).

### CODE REVIEW

File: src/porepy/applications/complete_setups/terzaghi_biot.py [hunk 1]
```diff
@@ -263,6 +266,12 @@ def consolidation_degree(self, t: number) -> float:
 class TerzaghiUtils(VerificationUtils):
     """Mixin class containing useful utility methods for the setup."""
 
+    applied_load: Callable[[], pp.number]
+    """Method that set the applied load in scaled [Pa]. Normally provided by an
```
Comment (IvarStefansson) [inline src/porepy/applications/complete_setups/terzaghi_biot.py:270]:
```suggestion
    """Method that sets the applied load in scaled [Pa]. Normally provided by an
```

File: src/porepy/applications/complete_setups/terzaghi_biot.py [hunk 2]
```diff
@@ -272,6 +281,12 @@ class TerzaghiUtils(VerificationUtils):
     bc_values_mechanics_key: str
     """Key for accessing mechanical boundary values."""
 
+    height: Callable[[], pp.number]
+    """Method that set the height of the domain in scaled [m]. Normally provided by an
```
Comment (IvarStefansson) [inline src/porepy/applications/complete_setups/terzaghi_biot.py:285]:
```suggestion
    """Method that sets the height of the domain in scaled [m]. Normally provided by an
```

File: src/porepy/applications/complete_setups/terzaghi_biot.py [hunk 3]
```diff
@@ -493,28 +520,36 @@ class PseudoOneDimensionalColumn(pp.ModelGeometry):
...
+        sd: pp.Grid = pp.CartGrid(n_cells, phys_dims)
+        sd.compute_geometry()
+
         self.domain = pp.Domain(
             {
-                "xmin": 0,
-                "xmax": phys_dims[0],
-                "ymin": 0,
-                "ymax": phys_dims[1],
+                "xmin": 0,  # scaled [m]
```
Comment (IvarStefansson) [inline src/porepy/applications/complete_setups/terzaghi_biot.py:542]: As a matter of taste, I think tagging "everything" with units is a bit too much.

File: src/porepy/applications/complete_setups/terzaghi_biot.py [hunk 4]
```diff
@@ -628,10 +670,16 @@ class TerzaghiPoromechanicsBoundaryConditions(
 class TerzaghiSolutionStrategy(poromechanics.SolutionStrategyPoromechanics):
     """Solution strategy class for Terzaghi's setup."""
 
+    applied_load: Callable[[], pp.number]
+    """Method that set the applied load in scaled [Pa]. Normally provided by an
```
Comment (IvarStefansson) [inline src/porepy/applications/complete_setups/terzaghi_biot.py:674]:
```suggestion
    """Method that sets the applied load in scaled [Pa]. Normally provided by an
```

File: tests/functional/test_terzaghi_biot.py [hunk 5]
```diff
@@ -131,3 +135,39 @@ def test_pressure_and_consolidation_degree_errors():
...
+    np.testing.assert_almost_equal(
+        unscaled.results[-1].error_pressure,
+        scaled.results[-1].error_pressure,
+        decimal=5,
+    )
+    np.testing.assert_almost_equal(
+        unscaled.results[-1].error_consolidation_degree,
+        unscaled.results[-1].error_consolidation_degree,
+        decimal=5,
+    )
```
Comment (IvarStefansson) [inline tests/functional/test_terzaghi_biot.py:173]: Missing newline, I think.

**Review by IvarStefansson — COMMENTED** [review]
Looks good. A couple of specific comments. Also, intuitively, I'm not sure if complete_setup is the correct place for setup_utils. Can you think of an alternative?

File: tests/functional/test_terzaghi_biot.py (hunk 5 above)
Comment (jhabriel) [reply tests/functional/test_terzaghi_biot.py:173]: Yes, fixed.

File: src/porepy/applications/building_blocks/models_glossary.py [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+    finalize_data_saving: Entry = Entry(
+        type="Callable[[], None]",
+        docstring="Finalize data saving. Normally provided by a mixin instance of"
+        " :class:`~porepy.viz.data_saving_model_mixin.DataSavingMixin`.",
+        name="finalize_data_saving",
+    )
+
+    fluid: Entry = Entry(
+        type="pp.FluidConstants",
+        docstring="Fluid constant object that takes care of scaling of fluid-related"
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/porepy/applications/building_blocks/models_glossary.py:276]: Fixed.

File: src/porepy/applications/building_blocks/models_glossary.py [hunk 7]
```diff
@@ -0,0 +1,737 @@
...
+        type="Callable[[list[pp.MortarGrid]], pp.ad.MixedDimensionalVariable]",
+        docstring="Variable for interface enthalpy flux. Normally provided by a mixin"
+        " instance of :class:`~porepy.models.energy_balance.VariablesEnergyBalance`.",
+        name="interface_enthalpy_flux",
+    )
+
+    interface_enthalpy_flux_variable: Entry = Entry(
+        type="str",
+        docstring="Name of the primary variable representing the enthalpy flux across"
+        " an interface. Normally defined in a mixin of instance"
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/porepy/applications/building_blocks/models_glossary.py:427]: Indeed. I will fixed them.

File: src/porepy/applications/building_blocks/models_glossary.py [hunk 8]
```diff
@@ -0,0 +1,737 @@
...
+    set_geometry: Entry = Entry(
+        type="Callable[[], None]",
+        docstring="Set the geometry of the model. Normally provided by a mixin instance"
+        " of :class:`~porepy.models.geometry.ModelGeometry`.",
+        name="set_geometry",
+    )
+
+    solid: Entry = Entry(
+        type="pp.SolidConstants",
+        docstring="Solid constant object that takes care of scaling of solid-related"
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/porepy/applications/building_blocks/models_glossary.py:587]: Done.

File: src/porepy/applications/building_blocks/models_glossary.py [hunk 9]
```diff
@@ -0,0 +1,737 @@
...
+        docstring="Fluid constant object that takes care of scaling of fluid-related"
+        " quantities. Normally, this is set by a mixin of instance"
+        " :class:`~porepy.models.solution_strategy.SolutionStrategy`.",
+        name="fluid",
+    )
+
+    fluid_density: Entry = Entry(
+        type="Callable[[list[pp.Grid]], pp.ad.Operator]",
+        docstring="Fluid density. Normally defined in a mixin class with a suitable"
+        " constitutive relation.",
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/porepy/applications/building_blocks/models_glossary.py:285]: Done. Also for other glossary entries that were vague.

File: src/porepy/applications/complete_setups/terzaghi_biot.py (hunk 3 above)
Comment (jhabriel) [reply src/porepy/applications/complete_setups/terzaghi_biot.py:542]: Sure. Fixed.

File: src/porepy/applications/building_blocks/models_glossary.py [hunk 10]
```diff
@@ -0,0 +1,760 @@
...
+    )
+
+    fluid_enthalpy: Entry = Entry(
+        type="Callable[[list[pp.Grid]], pp.ad.Operator]",
+        docstring="Variable for interface enthalpy flux. Normally provided by a mixin"
+        " instance of :class:`~porepy.models.energy_balance.VariablesEnergyBalance`.",
+        name="fluid_enthalpy",
+    )
+
+    # TODO: I could not find an usage of fluid_internal_energy in models
```
Comment (IvarStefansson) [inline src/porepy/applications/building_blocks/models_glossary.py:306]: It is used in the energy balance

File: src/porepy/applications/complete_setups/mandel_biot.py [hunk 11]
```diff
@@ -1284,7 +1245,10 @@ class MandelGeometry(pp.ModelGeometry):
     """Simulation model parameters."""
 
     fracture_network: pp.FractureNetwork2d
-    """Fracture network. Empty in this case."""
+    """Two-dimensional fracture network object. Set by a mixin instance of
```
Comment (IvarStefansson) [inline src/porepy/applications/complete_setups/mandel_biot.py:1248]: Are you sure this is needed for the class which assigns it?

File: src/porepy/applications/complete_setups/mandel_biot.py [hunk 12]
```diff
@@ -1314,24 +1278,32 @@ def set_md_grid(self) -> None:
...
+    instance of :class:`~porepy.models.geometry.ModelGeometry`.
+
+    """
 
     params: dict
     """Parameter dictionary of the verification setup."""
 
     stress_keyword: str
-    """Keyword for the mechanical subproblem."""
+    """Keyword for accesssing the parameters of the mechanical subproblem."""
```
Comment (IvarStefansson) [inline src/porepy/applications/complete_setups/mandel_biot.py:1300]: -s

File: src/porepy/numerics/fv/mpsa.py [hunk 13]
```diff
@@ -54,8 +54,8 @@ def __init__(self, keyword: str) -> None:
 
         self.stress_matrix_key = "stress"
         self.bound_stress_matrix_key = "bound_stress"
-        self.bound_displacment_cell_matrix_key = "bound_displacement_cell"
-        self.bound_displacment_face_matrix_key = "bound_displacement_face"
+        self.bound_displacement_cell_matrix_key = "bound_displacement_cell"
+        self.bound_displacement_face_matrix_key = "bound_displacement_face"
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/mpsa.py:58]: Haha!

**Review by IvarStefansson — COMMENTED** [review]
Almost there! Just need to talk to Eirik about VerificationUtils

File: src/porepy/applications/complete_setups/mandel_biot.py (hunk 11 above)
Comment (jhabriel) [reply src/porepy/applications/complete_setups/mandel_biot.py:1248]: Good catch. This is not needed. Fixed.

**Review by IvarStefansson — APPROVED** [review]

**Pull request merged**